llvm/lib/Transforms/Scalar/SROA.cpp

```diff
@@ -2257,6 +2257,38 @@ checkVectorTypesForPromotion(Partition &P, const DataLayout &DL,
   return nullptr;
 }
 
+static VectorType *createAndCheckVectorTypesForPromotion(
+    SetVector<Type *> &OtherTys,
+    SmallVectorImpl<VectorType *> &CandidateTysCopy,
+    function_ref<void(Type *)> CheckCandidateType, Partition &P,
+    const DataLayout &DL, SmallVector<VectorType *, 4> &CandidateTys,
+    bool &HaveCommonEltTy, Type *&CommonEltTy, bool &HaveVecPtrTy,
+    bool &HaveCommonVecPtrTy, VectorType *&CommonVecPtrTy) {
+  // Consider additional vector types where the element type size is a
+  // multiple of load/store element size.
+  for (Type *Ty : OtherTys) {
+    if (!VectorType::isValidElementType(Ty))
+      continue;
+    unsigned TypeSize = DL.getTypeSizeInBits(Ty).getFixedValue();
+    // Make a copy of CandidateTys and iterate through it, because we
+    // might append to CandidateTys in the loop.
+    for (VectorType *&VTy : CandidateTysCopy) {
```

Comment:
This isn't being mutated though? can you just use ArrayRef of CandidateTys?

Reply:
I think CheckCandidateType modifies CandidateTys

Reply:
But if it's append only the ArrayRef won't see the new elements

Reply:
For the purpose of the loop ArrayRef is fine I think since it is treated as fixed size, but we need the actual copy for the second call to `createAndCheckVectorTypesForPromotion` since we `CandidateTys.clear();`.

The clear() is done to save iterations in `checkVectorTypesForPromotion` by not retrying things, but we still need these Tys for the size-based `candidateTys` generation.

Reply:
Actually, this is causing a failure.

AFAICT, CandidateTys.push_back(VTy); under certain conditions is causing a change of the base address for CandidateTys. Then CandidateTysCopy points to garbage. I will change the PR

Reply:
Besides, we may `CandidateTys.clear()` from `CheckCandidateType`

Reply:
Add a test that hit this failure?

Reply:
There is no good way to capture the actual failure. This is because the failure depends on overwriting the reclaimed memory -- this is dependent upon malloc as well as hardware. 

Instead, I've added an assert which enforces memory invariance to support the test. This is one of the conditions for the failure.